packages/headless/src/controllers/commerce/instant-products/headless-instant-products.ts

```diff
@@ -164,6 +185,17 @@ export function buildInstantProducts(
       );
     },
 
+    promoteChildToParent(childPermanentId, parentPermanentId) {
```

Comment:
Why isn't the signature just `promoteChildToParent(childProduct: Product)` ?

Headless already knows the current "parent" (and thus it's permanent id)

Reply:
Mainly for performance reasons.

If we only pass the child permanent ID (or the full child product), we will have to traverse the children of each product in the controller's state in order to locate the child to promote (O(mn)).

Providing the parent ID in the call allows us to directly traverse the children of the parent product (O(n)).

Also, there wouldn't be an advantage to passing the full product: we only need the permanent id to locate it.

Reply:
> Also, there wouldn't be an advantage to passing the full product: we only need the permanent id to locate it.

The advantage would be for the implementer: Passing a `Product` interface is more prescriptive/restrictive (in a good way) than "any string will suffice as far as the compiler is concerned, but please use the one named permanent ID which is set as a property on a product". 

Even if internally we end up only needing permanentID to do the job.

> If we only pass the child permanent ID (or the full child product), we will have to traverse the children of each product in the controller's state in order to locate the child to promote (O(mn)).

It's again mostly for implementer POV.  It's more foolproof/harder to do a mistake if the onus is not on the implementer to pass 2 values "correctly linked to one another" (ie: the parent-child relationship is correctly respected by the implementer) VS having Headless handle it internally.

For atomic-product-children, the way the code is written, we have a solid relationship between the parent product and it's children (it's all colocated in the same file), so the advantage are minimal. 

But I can easily see a world where an implementer would build their own version of `atomic-product-children` with no reference to the parent product, or have each individual children product tile only have a reference to an individual child product.

And so you can imagine that the complexity would have to be shifted to the implementer.

For perf: We can add a reference to the parent product on each child product (in Headless reducer) to have it be a simple lookup.

Reply:
Will address in https://coveord.atlassian.net/browse/KIT-3221